firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java

```diff
@@ -112,14 +128,87 @@ public void onFailure(@NonNull @NotNull Exception e) {
     return taskCompletionSource.getTask();
   }
 
+  /** Signs in the App Distribution tester. Presents the tester with a Google sign in UI */
+  @NonNull
+  public Task<Void> signInTester() {
+    if (signInTaskCompletionSource != null && !signInTaskCompletionSource.getTask().isComplete()) {
+      signInCancellationSource.cancel();
+    }
+
+    signInCancellationSource = new CancellationTokenSource();
+    signInTaskCompletionSource = new TaskCompletionSource<>(signInCancellationSource.getToken());
+
+    Context context = firebaseApp.getApplicationContext();
+    AlertDialog alertDialog = getSignInAlertDialog(context);
+    alertDialog.show();
+
+    return signInTaskCompletionSource.getTask();
+  }
+
   /**
    * Returns an AppDistributionRelease if one is available for the current signed in tester. If no
    * update is found, returns null. If tester is not signed in, presents the tester with a Google
    * sign in UI
    */
   @NonNull
   public Task<AppDistributionRelease> checkForUpdate() {
-    return Tasks.forResult(null);
+
+    if (checkForUpdateTaskCompletionSource != null
+        && !checkForUpdateTaskCompletionSource.getTask().isComplete()) {
+      checkForUpdateCancellationSource.cancel();
+    }
+
+    checkForUpdateCancellationSource = new CancellationTokenSource();
+    checkForUpdateTaskCompletionSource =
+        new TaskCompletionSource<>(checkForUpdateCancellationSource.getToken());
+
+    firebaseInstallationsApi
+        .getId()
+        .addOnSuccessListener(
+            new OnSuccessListener<String>() {
+              @Override
+              public void onSuccess(String fid) {
+                firebaseInstallationsApi
+                    .getToken(true)
+                    .addOnSuccessListener(
```

Comment:
This block of code is deeply nested. Could we extract out the tasks to make this a little cleaner?

    Task<String> installationIdTask = firebaseInstallations.getId();
    Task<InstallationTokenResult> installationTokenTask = firebaseInstallations.getToken(true);

Then I think we can do something fancy with the Tasks API:
Tasks.whenAllSuccess(installationIdTask, installationTokenTask).addOnSuccessListener( 
 new OnSuccessListener<List<Object>>() {
          @Override
          public void onSuccess(List<Object> objects) {
            String fid = installationIdTask.getResult();
            String token = installationTokenTask.getResult();
          }


Let me know if you want to chat through this idea!

Reply:
I like it, very clean

Reply:
Nice!